dpdk/vr_dpdk_lcore.c

```diff
@@ -363,76 +461,58 @@ vr_dpdk_packets_vroute(struct vr_interface *vif, struct vr_packet *pkts[VR_DPDK_
+    uint64_t total_pkts = 0;
+    struct rte_mbuf *pkts[VR_DPDK_MAX_BURST_SZ];
+    struct vr_dpdk_queue *rx_queue;
+    uint32_t nb_pkts;
     struct vr_packet *pkt_arr[VR_DPDK_MAX_BURST_SZ];
     int pkti;
 
     /* for all RX queues */
-    while (cur_queues_mask) {
-        rte_prefetch0(rx_queue);
-        if (likely(cur_queues_mask & cur_queue)) {
-            /* burst RX */
-            nb_pkts = rx_queue->rxq_ops.f_rx(rx_queue->rxq_queue_h, pkts,
+    SLIST_FOREACH(rx_queue, &lcore->lcore_rx_head, q_next) {
+        /* burst RX */
+        nb_pkts = rx_queue->rxq_ops.f_rx(rx_queue->q_queue_h, pkts,
                 rx_queue->rxq_burst_size);
-            cur_queues_mask &= ~cur_queue;
-            if (likely(nb_pkts > 0)) {
-                /* transmit packets to vrouter */
-                if (vif_is_virtual(rx_queue->rxq_vif)) {
-                    for (pkti = 0; pkti < nb_pkts; pkti++) {
-                        pkt_arr[pkti] = vr_dpdk_packet_get(pkts[pkti],
-                                                           rx_queue->rxq_vif);
-                    }
-                    vr_dpdk_virtio_enq_pkts_to_phys_lcore(rx_queue,
-                                                          pkt_arr, nb_pkts);
-                } else {
-                    dpdk_vroute(rx_queue->rxq_vif, pkts, nb_pkts);
+        if (likely(nb_pkts > 0)) {
+            total_pkts += nb_pkts;
+            /* transmit packets to vrouter */
+            if (vif_is_virtual(rx_queue->q_vif)) {
+                for (pkti = 0; pkti < nb_pkts; pkti++) {
+                    pkt_arr[pkti] = vr_dpdk_packet_get(pkts[pkti],
+                                                       rx_queue->q_vif);
                 }
+                vr_dpdk_virtio_enq_pkts_to_phys_lcore(rx_queue,
+                                                      pkt_arr, nb_pkts);
             } else {
-                /* mark the queue as empty */
-                *rx_queues_mask &= ~cur_queue;
+                dpdk_vroute(rx_queue->q_vif, pkts, nb_pkts);
             }
         }
-        cur_queue <<= 1;
-        rx_queue++;
     }
-    return nb_pkts;
+    return total_pkts;
 }
 
 /* Forwarding lcore IO */
 static inline void
 dpdk_lcore_fwd_io(struct vr_dpdk_lcore *lcore)
 {
-    uint64_t rx_queues_mask = lcore->lcore_rx_queues_mask;
     uint64_t total_pkts = 0;
     struct rte_mbuf *pkts[VR_DPDK_MAX_BURST_SZ];
     uint32_t nb_pkts;
     int i;
     struct vr_dpdk_ring_to_push *rtp;
-    uint16_t num_tx_rings, max_tx_rings;
+    uint16_t nb_rtp;
 
-    total_pkts += dpdk_lcore_fwd_rx(lcore, &pkts[0], &rx_queues_mask);
-    total_pkts += dpdk_lcore_fwd_rx(lcore, &pkts[0], &rx_queues_mask);
-    total_pkts += dpdk_lcore_fwd_rx(lcore, &pkts[0], &rx_queues_mask);
-    total_pkts += dpdk_lcore_fwd_rx(lcore, &pkts[0], &rx_queues_mask);
-    total_pkts += dpdk_lcore_fwd_rx(lcore, &pkts[0], &rx_queues_mask);
+    total_pkts += dpdk_lcore_fwd_rx(lcore);
+    total_pkts += dpdk_lcore_fwd_rx(lcore);
```

Comment:
RX operation for KNIs is quite expensive. We used rx_queue_mask to mask out the ports with no packets to read (i.e. read them less frequently). IMO we need to implement the same functionality for the list of RX queues now.

I added the TODO comment.